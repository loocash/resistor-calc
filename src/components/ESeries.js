import React from 'react';
import PropTypes from 'prop-types';
import { stripZero } from "../lib/utils";

import Table from 'react-bootstrap/Table';
//import Badge from 'react-bootstrap/Badge';

import {base} from "../lib/utils";
import { bandsToESeries, eseriesToTolerances, getTriple } from "../lib/eseries";

const ResistanceLink = ({base, onBaseChange}) =>
    <a onClick={(e) => {
      e.preventDefault();
      onBaseChange(base);
    }}
       href="#">{base}</a>;

const ToleranceLink = ({tolerance, onToleranceChange}) =>
    <a onClick={(e) => {
      e.preventDefault();
      onToleranceChange(tolerance);
    }}
       href="#">{stripZero(tolerance)}</a>;

/*const QuestionBadge = () => (
  <Badge pill variant="info"
    style={{
      display: 'inline',
      lineHeight: 2,
      verticalAlign: 'top',
      cursor: 'pointer',
    }}
  >
    ?
  </Badge>
);*/


export default function ESeries({
  resistor,
  onBaseChange,
  onToleranceChange,
}) {
  const {resistance, tolerance, bands} = resistor;
  const baseResistance = base(resistance);
  const eseries = bandsToESeries[bands];

  return (
    <div className="ESeries">

      <Table striped hover size="sm">
        <thead>
          <tr>
            <th>E-Series {/*<QuestionBadge />*/}</th>
            <th>Tolerances [%]</th>
            <th>Less</th>
            <th>Equal</th>
            <th>Greater</th>
          </tr>
        </thead>
        <tbody>
        {eseries.map(series => {
          const [smaller, equal, greater] = getTriple(baseResistance, series);
          return (
            <tr key={series}>
              <td>{series}</td>
              <td>{eseriesToTolerances[series].map((tol, index) => (
                <React.Fragment key={index}>
                  {tolerance !== tol ? <ToleranceLink tolerance={tol} onToleranceChange={onToleranceChange}/> :
                    stripZero(tolerance)}
                  {index < eseriesToTolerances[series].length - 1 ? ', ' : ''}
                </React.Fragment>
              ))}
              </td>
              <td>{smaller ? <ResistanceLink base={smaller} onBaseChange={onBaseChange} /> : '—'}</td>
              <td>{equal ? equal : '—'}</td>
              <td>{greater ? <ResistanceLink base={greater} onBaseChange={onBaseChange} /> : '—'}</td>
            </tr>
          );
        })}
        </tbody>
      </Table>
    </div>
  );
}

ESeries.propTypes = {
  resistor: PropTypes.object.isRequired,
  onBaseChange: PropTypes.func.isRequired,
  onToleranceChange: PropTypes.func.isRequired,
};
