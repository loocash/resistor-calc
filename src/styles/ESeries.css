.ESeries span {
    vertical-align: middle;
}

.ESeries table {
    width: 100%;
    text-align: center;
}
