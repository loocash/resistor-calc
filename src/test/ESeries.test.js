import React from 'react';
import ESeries, { getTriple, baseResistance } from '../components/ESeries';
import { render, cleanup, fireEvent } from '@testing-library/react';

beforeEach(cleanup);

test('calls onBaseChange when a new base is clicked', () => {
  const props = {
    resistance: 150,
    onBaseChange: jest.fn(),
    onToleranceChange: jest.fn(),
    bands: 4,
  };
  const { getByText } = render(<ESeries {...props} />);
  const base13 = getByText('1.3');
  fireEvent.click(base13);
  expect(props.onBaseChange)
    .toHaveBeenCalledWith(1.3);
});

[{
  resistance: 931,
  bands: 5,
}, {
  resistance: 100,
  bands: 4,
},
].forEach(({resistance, bands}) => {
  test(`matches snapshot for resistance: ${resistance}`, () => {
    const { container } = render(
      <ESeries resistance={resistance}
               bands={bands}
               onBaseChange={jest.fn()}
               onToleranceChange={jest.fn()}
      />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

[{
  base: 9.53,
  series: 'E96',
  want: [9.31, 9.53, 9.76],
}, {
  base: 1,
  series: 'E6',
  want: [null, 1, 1.5],
}, {
  base: 3.48,
  series: 'E192',
  want: [3.44, 3.48, 3.52],
},
].forEach(({base, series, want}) => {
  test(`getTriple(${base}, ${series}) should return: ${want}`, () => {
    const got = getTriple(base, series);
    expect(got).toEqual(want);
  });
});

[{
  resistance: 100,
  want: 1,
}, {
  resistance: 348,
  want: 3.48,
},
].forEach(({resistance, want}) => {
  test(`baseResistance(${resistance}) should return ${want}`, () => {
    const got = baseResistance(resistance);
    expect(got).toBe(want);
  });
});
