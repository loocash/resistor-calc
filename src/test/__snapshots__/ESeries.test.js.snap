// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for resistance: 100 1`] = `
<div
  class="ESeries"
>
  <table
    class="table table-sm table-striped table-hover"
  >
    <thead>
      <tr>
        <th>
          E-Series 
          <span
            class="badge badge-pill badge-info"
            style="display: inline; line-height: 2; vertical-align: top; cursor: pointer;"
          >
            ?
          </span>
        </th>
        <th>
          Tolerances [%]
        </th>
        <th>
          Less
        </th>
        <th>
          Equal
        </th>
        <th>
          Greater
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          E12
        </td>
        <td>
          <a
            href="#"
          >
            10
          </a>
          
        </td>
        <td>
          —
        </td>
        <td>
          1
        </td>
        <td>
          <a
            href="#"
          >
            1.2
          </a>
        </td>
      </tr>
      <tr>
        <td>
          E24
        </td>
        <td>
          <a
            href="#"
          >
            5
          </a>
          
        </td>
        <td>
          —
        </td>
        <td>
          1
        </td>
        <td>
          <a
            href="#"
          >
            1.1
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`matches snapshot for resistance: 931 1`] = `
<div
  class="ESeries"
>
  <table
    class="table table-sm table-striped table-hover"
  >
    <thead>
      <tr>
        <th>
          E-Series 
          <span
            class="badge badge-pill badge-info"
            style="display: inline; line-height: 2; vertical-align: top; cursor: pointer;"
          >
            ?
          </span>
        </th>
        <th>
          Tolerances [%]
        </th>
        <th>
          Less
        </th>
        <th>
          Equal
        </th>
        <th>
          Greater
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          E48
        </td>
        <td>
          <a
            href="#"
          >
            2
          </a>
          
        </td>
        <td>
          <a
            href="#"
          >
            9.09
          </a>
        </td>
        <td>
          —
        </td>
        <td>
          <a
            href="#"
          >
            9.53
          </a>
        </td>
      </tr>
      <tr>
        <td>
          E96
        </td>
        <td>
          <a
            href="#"
          >
            1
          </a>
          
        </td>
        <td>
          <a
            href="#"
          >
            9.09
          </a>
        </td>
        <td>
          9.31
        </td>
        <td>
          <a
            href="#"
          >
            9.53
          </a>
        </td>
      </tr>
      <tr>
        <td>
          E192
        </td>
        <td>
          <a
            href="#"
          >
            .5
          </a>
          , 
          <a
            href="#"
          >
            .25
          </a>
          , 
          <a
            href="#"
          >
            .1
          </a>
          , 
          <a
            href="#"
          >
            .05
          </a>
          
        </td>
        <td>
          <a
            href="#"
          >
            9.2
          </a>
        </td>
        <td>
          9.31
        </td>
        <td>
          <a
            href="#"
          >
            9.42
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
